### PR TITLE
OUTDATED: change(state): Add note subtree indexes to the database, and calculate subtrees for existing blocks

### LINKED ISSUE #6953: Add note subtree indexes to the finalized and non-finalized state
## Motivation

We want to create a new database column family to support the `z_getsubtreesbyindex` RPC.

## Specifications

See the [state upgrade design docs](https://github.com/ZcashFoundation/zebra/blob/main/book/src/dev/state-db-upgrades.md#design).

`zcashd` implemented support for rebuilding and regenerating their subtree database based on existing blocks during initialization. We may not want to reproduce this exactly but it may be informative: https://github.com/zcash/zcash/pull/6694

### State Format Design

- We'll need a new column family that maps:
    - `subtree_index` to `(end_height, subtree_root)`

### Calculating subtree indexes

We need to calculate fixed-size note commitment subtrees for existing and new blocks.

Here is an algorithm and examples of the calculation:

<details>

<summary>implementation details</summary>

`zcashd`'s best chain / subtrees state PR:
https://github.com/zcash/zcash/pull/6649

In particular:
https://github.com/zcash/zcash/pull/6649/files#diff-31ef5cdda38b86238aa5fa0745b7843ebd43f11d65583fb9cfec23f7225e5f77

For each subtree in the history of the note commitment tree (Sapling, Orchard):
- every time a block that completes a subtree is appended to the chain, we will store a new subtree (a block can only complete one subtree due to block size limits)
    - some blocks will complete a subtree and start a new one
    - some blocks will complete a subtree, and the next block will start a new one
- if a reorg happens (not relevant to Zebra)
- light clients need to be able to fetch any subtrees from `current_height - 100` onwards, using bitwise math (still to be specified)
    - this will probably require scanning through the database for trees, or building an index of note commitment counts per block

- We can calculate the subtree roots efficiently using: https://github.com/zcash/zcash/pull/6431/commits/1e23185ffa3b3e623f588524aa8bf3d4cd4abbb2#diff-dd19eaef96b055fb2c94eaa1a7bcda8074fb912e2fefc0695e0b4c05a76d533eR121-L108 see the discussion at https://discord.com/channels/809218587167293450/809251012610752513/1110592469575422016

</details>

### Complex Code or Requirements

State lookups are concurrent, so we'll need to deal with concurrent database updates where the database overlaps the cloned non-finalized state.

## Implementation Steps

 - [ ] update the [database format](https://github.com/ZcashFoundation/zebra/blob/main/book/src/dev/state-db-upgrades.md#current) in the Zebra docs
 - [ ] increment the state minor version
 - [ ] write the new format in the block write task
 - [ ] update older formats in the format upgrade task

### Testing

Database format changes will need:
 - [ ] test that the new format works when creating a new state, and updating an older state

New state queries will need:
- [ ] new unit tests
- [ ] new snapshot tests (at the zebra_db.rs API level)

And possibly:
- [ ] new integration tests

### PR DESCRIPTION
**Pull request by arya2**
## Motivation

This is needed to support "spend before sync" in lightwalletd.

Closes #6953.

Depends-On #7334.

TODO: Add state validity test.

## Solution

- Add subtrees below initial tip height

## Review

Not yet ready for review.

### Reviewer Checklist

  - [ ] Will the PR name make sense to users?
    - [ ] Does it need extra CHANGELOG info? (new features, breaking changes, large changes)
  - [ ] Are the PR labels correct?
  - [ ] Does the code do what the ticket and PR says?
    - [ ] Does it change concurrent code, unsafe code, or consensus rules?
  - [ ] How do you know it works? Does it have tests?

## Follow Up Work

- Add state service requests for reading subtrees

### CODE REVIEW

File: zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+                    )) {
+                        panic!("unexpected send error: {error:?}");
+                    }
+                }
+
+                if subtree_index < tree_subtree_index {
+                    start_height = end_height;
+                    subtree_index = tree_subtree_index;
+                }
+            }
```
Comment (arya2) [inline zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs]: This is wrong.

File: zebra-state/src/service/finalized_state/disk_format/upgrade.rs [hunk 2]
```diff
@@ -74,6 +78,31 @@ pub struct DbFormatChangeThreadHandle {
...
+        version
+            .parse()
+            .expect("Hardcoded version string should be valid.")
+    }
+
+    vec![
+        (parse_expect_version("25.1.0"), Box::new(PruneTrees)),
+        (parse_expect_version("25.2.0"), Box::new(AddSubtrees)),
+    ]
+}
```
Comment (teor2345) [inline zebra-state/src/service/finalized_state/disk_format/upgrade.rs]: Turning the upgrades into a list is a good refactor, but large code movements should go in a separate PR, without any changes to functionality.

That makes them much easier to review and test.

File: zebra-state/src/service/finalized_state/disk_format/upgrade.rs [hunk 3]
```diff
@@ -74,6 +78,31 @@ pub struct DbFormatChangeThreadHandle {
...
+fn format_upgrades() -> Vec<(Version, Box<dyn DiskFormatUpgrade>)> {
+    fn parse_expect_version(version: &str) -> Version {
+        version
+            .parse()
+            .expect("Hardcoded version string should be valid.")
+    }
+
+    vec![
+        (parse_expect_version("25.1.0"), Box::new(PruneTrees)),
+        (parse_expect_version("25.2.0"), Box::new(AddSubtrees)),
```
Comment (teor2345) [inline zebra-state/src/service/finalized_state/disk_format/upgrade.rs]: Since the `25.2.0` version change is going to be merged in a previous PR, and merging that PR will run CI, we can't re-use that version in this PR. Instead, finish any changes to the format in this PR, and then increment the version again.

Normally we could delete all of the `25.2` cached states in Google Cloud, but I don't think that will work here, because it's a manual step that we can't make Mergify do. (And if we don't do that manual step before each CI run, the upgrade code in this PR will be skipped during the tests.)

```suggestion
        (parse_expect_version("25.3.0"), Box::new(AddSubtrees)),
```

File: zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs [hunk 4]
```diff
@@ -0,0 +1,181 @@
+//! Adds Sapling and Orchard note commitment subtrees to database
```
Comment (teor2345) [inline zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs]: Nit: The previous PR adds the subtrees, this PR fully populates them

```suggestion
//! Fully populate the Sapling and Orchard note commitment subtrees for existing blocks in the database.
```

File: zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs (hunk 1 above)
Comment (teor2345) [reply zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs]: There's an "implementation details" section of ticket #6953 that might be helpful here, the details are currently behind a disclosure triangle.
Screenshot 2023-08-21 at 14 24 30

File: zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+                }
+
+                if let Err(error) = sapling_tree_tx.send((height, tree)) {
+                    panic!("unexpected send error: {error:?}");
+                }
+            }
+        });
+
+        // Create an unbounded channel for sending sapling note commitment subtrees to be written
+        let (sapling_subtree_tx, sapling_subtree_rx) = mpsc::channel();
```
Comment (teor2345) [inline zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs]: Overall, I feel like this design is more complicated than it needs to be. Is it necessary to launch separate threads to do this upgrade?

It could be easier to implement it without threads, and then see if the performance is acceptable. (Remember, the upgrade task itself is already running in a separate thread.)

#### Security / Reliability

Using an unbounded channel can potentially read all the trees into RAM, if the write thread happens to block. How much memory is that? If it's a large amount, this is a potential hang or crash issue.

(I think it should be ok after duplicates are removed, and the duplicate removal task itself is ok because its compare task drops most of those trees immediately.)

File: zebra-state/src/service/finalized_state/disk_format/upgrade.rs (hunk 3 above)
Comment (arya2) [reply zebra-state/src/service/finalized_state/disk_format/upgrade.rs]: > Since the 25.2.0 version change is going to be merged in a previous PR, and merging that PR will run CI, we can't re-use that version in this PR. Instead, finish any changes to the format in this PR, and then increment the version again.

The previous PR shouldn't merge into main without this first merging into it.

File: zebra-state/src/service/finalized_state/disk_format/upgrade.rs (hunk 2 above)
Comment (teor2345) [reply zebra-state/src/service/finalized_state/disk_format/upgrade.rs]: Actually, I think it might be better to delay this refactor for a while, because it will cause merge conflicts with bug fixes to the `25.1` state upgrade.

File: zebra-state/src/service/finalized_state/disk_format/upgrade.rs (hunk 3 above)
Comment (teor2345) [reply zebra-state/src/service/finalized_state/disk_format/upgrade.rs]: Ah, then we'll need to delete all the `25.2` states before doing the final tests and merging both PRs.

In fact, we might need to do that deletion on a regular basis, to get decent CI results from both PRs.

File: zebra-state/src/service/finalized_state/disk_format/upgrade.rs (hunk 2 above)
Comment (arya2) [reply zebra-state/src/service/finalized_state/disk_format/upgrade.rs]: > Actually, I think it might be better to delay this refactor for a while, because it will cause merge conflicts with bug fixes to the 25.1 state upgrade.

I think it could help with avoiding merge conflicts if we do the upgrade for `25.2.1` in parallel to `25.1.1`.

Comment (teor2345) [reply zebra-state/src/service/finalized_state/disk_format/upgrade.rs]: > > Actually, I think it might be better to delay this refactor for a while, because it will cause merge conflicts with bug fixes to the 25.1 state upgrade.
> 
> I think it could help with avoiding merge conflicts if we do the upgrade for `25.2.1` in parallel to `25.1.1`.

I'm not sure what you mean by "in parallel"". Do you mean "concurrently running in a separate thread", "in the same PR", or "in the same release"?

But it's not just merge conflicts that are a factor here, it's technical complexity risk, reviewer attention, testing, and developer time/scope. Reducing code changes helps with all of these.

File: zebra-state/src/constants.rs [hunk 6]
```diff
@@ -48,7 +48,7 @@ pub(crate) const DATABASE_FORMAT_VERSION: u64 = 25;
 /// - adding new column families,
 /// - changing the format of a column family in a compatible way, or
 /// - breaking changes with compatibility code in all supported Zebra versions.
-pub(crate) const DATABASE_FORMAT_MINOR_VERSION: u64 = 1;
+pub(crate) const DATABASE_FORMAT_MINOR_VERSION: u64 = 2;
```
Comment (teor2345) [inline zebra-state/src/constants.rs:51]: Note that the patch version will also need to be reset to zero.

File: zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+/// Runs disk format upgrade for adding Sapling and Orchard note commitment subtrees to database.
+pub fn _run(
+    initial_tip_height: Height,
+    upgrade_db: &ZebraDb,
+    should_cancel_format_change: Arc<AtomicBool>,
+) {
+    let mut subtree_count = 0;
+    let mut prev_tree: Option<_> = None;
+    for (height, tree) in upgrade_db.sapling_tree_by_height_range(..=initial_tip_height) {
```
Comment (teor2345) [inline zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs:23]: Note: I split the lightwalletd back-tracking requirement into ticket #7407.

(Please don't fix it in this PR, it needs its own code and tests, and this PR is big enough already.)

File: zebra-state/src/service/finalized_state/zebra_db/shielded.rs [hunk 8]
```diff
@@ -355,9 +396,30 @@ impl DiskWriteBatch {
             self.zs_insert(&orchard_tree_cf, height, trees.orchard);
         }
 
+        if let Some(subtree) = trees.sapling_subtree {
+            self.zs_insert(&sapling_subtree_cf, subtree.index, subtree.into_data());
```
Comment (teor2345) [inline zebra-state/src/service/finalized_state/zebra_db/shielded.rs:400]: Note: Since this PR writes to the database, it can't merge until after the 1.2.0 release.

File: zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+        if should_cancel_format_change.load(atomic::Ordering::Relaxed) {
+            break;
+        }
```
Comment (teor2345) [inline zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs:26]: Note: the cancel handle got reverted to the channel implementation, because atomics don't correctly handle dropped states.

File: zebra-state/src/service/finalized_state/disk_format/tests/prop.rs [hunk 10]
```diff
@@ -361,6 +362,16 @@ fn roundtrip_sapling_tree_root() {
     proptest!(|(val in any::<sapling::tree::Root>())| assert_value_properties(val));
 }
 
+#[test]
+fn roundtrip_sapling_subtree_data() {
+    let _init_guard = zebra_test::init();
+
+    proptest!(|(mut val in any::<NoteCommitmentSubtreeData<sapling::tree::Node>>())| {
+        val.end = val.end.clamp(Height(0), MAX_ON_DISK_HEIGHT);
```
Comment (teor2345) [inline zebra-state/src/service/finalized_state/disk_format/tests/prop.rs:370]: This might not do what you expect, because almost all the tests will be checking an `end` of `MAX_ON_DISK_HEIGHT`.

This issue is documented here, but for a string regex, rather than a clamp:
> In general, the ideal solution is to find a way to take a seed value and then use prop_map to transform it into the desired, irregular domain.
> This kind of works, but it has problems ... [it] also results in a pretty odd distribution of values. 

https://proptest-rs.github.io/proptest/proptest/tutorial/filtering.html
https://proptest-rs.github.io/proptest/proptest/tutorial/transforming-strategies.html

Instead, transform the input range so its distribution is roughly uniform:
```suggestion
        val.end.0 %= (MAX_ON_DISK_HEIGHT.0 + 1);
```

File: zebra-state/src/service/finalized_state/disk_format/tests/prop.rs [hunk 11]
```diff
@@ -436,6 +447,16 @@ fn roundtrip_orchard_tree_root() {
     proptest!(|(val in any::<orchard::tree::Root>())| assert_value_properties(val));
 }
 
+#[test]
+fn roundtrip_orchard_subtree_data() {
+    let _init_guard = zebra_test::init();
+
+    proptest!(|(mut val in any::<NoteCommitmentSubtreeData<orchard::tree::Node>>())| {
+        val.end = val.end.clamp(Height(0), MAX_ON_DISK_HEIGHT);
```
Comment (teor2345) [inline zebra-state/src/service/finalized_state/disk_format/tests/prop.rs:455]:
```suggestion
        val.end.0 %= (MAX_ON_DISK_HEIGHT.0 + 1);
```

File: zebra-state/src/service/finalized_state/zebra_db/shielded.rs [hunk 12]
```diff
@@ -224,7 +260,9 @@ impl ZebraDb {
         NoteCommitmentTrees {
             sprout: self.sprout_tree(),
             sapling: self.sapling_tree(),
+            sapling_subtree: None,
             orchard: self.orchard_tree(),
+            orchard_subtree: None,
```
Comment (teor2345) [inline zebra-state/src/service/finalized_state/zebra_db/shielded.rs:265]: Looks like you might need `*_subtree_for_tip()` methods here.

You can get the tip trees with `zs_last_key_value()`, feel free to grab the `impl FromDisk for NoteCommitmentSubtreeIndex` from PR #7408.

**Comment by teor2345** [comment]
I had to change more code than I expected in the RPC PR, so if you could minimise the changes in this PR, that would be helpful!

Then we can merge them both, and base any fixes on the `main` branch.

File: zebra-state/src/constants.rs (hunk 6 above)
Comment (teor2345) [reply zebra-state/src/constants.rs:51]: I fixed this in PR #7437.

File: zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs (hunk 7 above)
Comment (teor2345) [reply zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs:23]: I changed the code below this code in PR #7437, so fixing it here would cause merge conflicts.

File: zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs (hunk 9 above)
Comment (teor2345) [reply zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs:26]: I fixed this in PR #7437.

File: zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs (hunk 7 above)
Comment (teor2345) [reply zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs:23]: Let's continue work on this in the ticket.

**Comment by teor2345** [comment]
@arya2 I needed the data from this PR to test the RPC method, so I copied that code across into PR #7437, and did the necessary fixes.

Let's keep working on the database changes in that PR?

I suggested some minor fixes to a test and a production method in this PR. Can you check if anything needs to be done?
Feel free to open new PRs or tickets for those fixes.

File: zebra-state/src/service/finalized_state/disk_format/tests/prop.rs (hunk 10 above)
Comment (teor2345) [reply zebra-state/src/service/finalized_state/disk_format/tests/prop.rs:370]: This is now #7443

File: zebra-state/src/service/finalized_state/zebra_db/shielded.rs (hunk 12 above)
Comment (teor2345) [reply zebra-state/src/service/finalized_state/zebra_db/shielded.rs:265]: This is now ticket #7444
